File: 2021W26/B0neh3ad/pgr_77485.py
def rotation(matrix, q):
    y1, x1, y2, x2 = q
    temp = matrix[y1 - 1][x2 - 1]
    min_val = temp

    # push top 1 block right
    matrix[y1 - 1][x1:x2] = matrix[y1 - 1][x1 - 1:x2 - 1]
    min_val = min(matrix[y1 - 1][x1:x2] + [min_val])

    # pull left 1 block up
    for i in range(y1, y2):
        matrix[i - 1][x1 - 1] = matrix[i][x1 - 1]
        min_val = min(matrix[i - 1][x1 - 1], min_val)

    # push bottom 1 block left
    matrix[y2 - 1][x1 - 1:x2 - 1] = matrix[y2 - 1][x1:x2]
    min_val = min(matrix[y2 - 1][x1 - 1:x2 - 1] + [min_val])

    # push right 1 block down
    for i in range(y2 - 1, y1 - 1, -1):
        matrix[i][x2 - 1] = matrix[i - 1][x2 - 1]
        min_val = min(matrix[i][x2 - 1], min_val)

    # finish rotation
    matrix[y1][x2 - 1] = temp

    return min_val


def solution(rows, columns, queries):
    answer = []
    numbers = [list(range(columns*i + 1, columns*(i+1) + 1)) for i in range(rows)]

    for query in queries:
        answer.append(rotation(numbers, query))

    return answer

File: 2021W26/fienestar/pgr_77485.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;
    answer.reserve(queries.size());
    
    vector<vector<int>> board(rows, vector<int>(columns));
    {
        size_t k = 1;
        for(auto& line:board)
            for(auto& x:line)
                x = k++;
    }
    
    auto rotate = [&](size_t x1, size_t y1, size_t x2, size_t y2) {
        static int dx[] = {0,1,0,-1};
        static int dy[] = {1,0,-1,0};

        auto x = x1;
        auto y = y1;

        int last = board[x][y];
        int m = board[x][y];
        for(size_t k=0; k!=4; ++k)
            while(x + dx[k] - x1 < x2 - x1 && y + dy[k] - y1 < y2 - y1){
                x = x + dx[k];
                y = y + dy[k];
                
                m = min(m, board[x][y]);
                
                swap(last, board[x][y]);
            }
        answer.push_back(m);
    };
    
    for(auto& query:queries)
        rotate(query[0]-1, query[1]-1, query[2], query[3]);
    
    return answer;
}